compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -819,24 +814,27 @@ public void visit(BLangMatchStatement matchStatement) {
             }
             this.resetErrorThrown();
             analyzeNode(matchClause, env);
-            allClausesReturns = allClausesReturns && this.matchClauseReturns;
-            containsLastPatternInStatement =
-                    containsLastPatternInStatement || (matchClause.matchGuard == null && this.hasLastPatternInClause);
+            allClausesReturns = allClausesReturns && this.statementReturns;
+            resetStatementReturns();
+            this.hasLastPatternInStatement =
+                    this.hasLastPatternInStatement || (matchClause.matchGuard == null && this.hasLastPatternInClause);
```

Comment:
We can move this logic to the visit of `BLangMatchClause`, right? We can probably get rid of the `hasLastPatternInClause` global variable by doing so.

Reply:
Fixed